### DISCUSSION TITLE
IntelliSense in Visual Studio Code

### DISCUSSION THREAD

**Discussion by Gonta01** [nasa/fprime#2807] [Q&A] [answered]
Does somebody know how to properly configure F' project, so that I can properly navigate through it without red line warnings of VSCode? What extensions should be used for it? Cmake, C/C++ externsion?

**Comment by thomas-bc** [comment, marked as answer]
I have been messing with VSCode settings for a while, so figured I would do a little config-dump here

## Useful extensions
- C/C++ extension pack: https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools-extension-pack
- CMake Tools https://marketplace.visualstudio.com/items?itemName=ms-vscode.cmake-tools
- FPP Language support https://marketplace.visualstudio.com/items?itemName=jet-propulsion-laboratory.fpp
- For FPP development, Scala (Metals): https://marketplace.visualstudio.com/items?itemName=scalameta.metals

## IntelliSense 
In `settings.json` (either at User level, or within your workspace):
```
    "C_Cpp.default.configurationProvider": "ms-vscode.cmake-tools",
    "C_Cpp.default.includePath": [
        "${workspaceFolder}/fprime",
        "${workspaceFolder}/**"
    ]
```
This can be made more efficient by pulling the actual list of include paths from the CMake cache, but `**` was ok for me. It is slow (few seconds) at first, but then caches the results so navigation is seamless.

## Debug F´ App with CMake extension
This was tricky to set up, but allows to debug F´ C++ in VSCode. The issue is that VSCode by default run in a fresh environment to run the CMake build steps. However we need an F´ environment to have access to the FPP tools etc. Therefore we have to add configuration for CMake to know where to find them. I have in the below example used a conda environment so I can isolate it further, but likely a Python venv should be sufficient too.
In `settings.json`:
```
    "cmake.buildDirectory": "${workspaceFolder}/build-fprime-vscode",
    "cmake.installPrefix": "${workspaceFolder}/build-artifacts/vscode",
    "cmake.debugConfig": {
        "args": ["-p", "50000", "-a", "127.0.0.1"] // args to pass to the FSW executable
    },
    "cmake.environment": {
        "PATH": "/path/to/.micromamba/envs/vscode-cmake-debug/bin:/path/to/.conda/condabin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin",
        "PYTHONPATH": "/path/to/.conda/envs/vscode-cmake-debug/lib/"
    },
```

Then in the CMake tab of VSCode, you can search for your deployment's target, and build/run/debug it! 🎉 

## Debug FPP / Scala code
Caveats: need to open `fpp/compilers` in VScode rather than `fpp/` (top level), as it is the root of the Scala project. There may be a configuration to change the root, but I have not found it.
Then, install Scala/Metals extension, and use a configuration like the following:

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "scala",
            "request": "launch",
            "name": "Launch FppToDict",
            // full name of the class to run
            "mainClass": "fpp.compiler.FPPToDict",
            // Args will need extra arguments / sources / includes. Run `fprime-util build --verbose` to get them printed out
            "args": ["-h"],
            // optional jvm properties to use
            "jvmOptions": []
        },
   ]
}
```

**Reply by Gonta01** [reply]
@thomas-bc thank you for your response!
Could you give more details about **Debug F´ App with CMake extension**. Don't understand what paths I should set for **cmake.environment**

**Reply by thomas-bc** [reply]
You basically need to specify a `PATH` environment variable that will have all the tools F´ needs to build properly. Maybe something that can work is
```
cd my-project
source fprime-venv/bin/activate
echo $PATH
```

Similarly with PYTHONPATH, you'll need the `lib/` folder of your virtual environment in there

**Reply by thomas-bc** [reply]
Btw this is only needed if you are in a machine where you install pip packages in virtual environments and stuff. If you are in an environment where packages are installed globally (e.g. a Docker container with the VSCode SSH thing), you should be good to go.

**Comment by LeStarch** [comment]
If you follow the standard F´  project format (virtual environment named `fprime-venv` in the project root), then IDE integration is fairly straightforward using `CMakePresets.json`.  You choose your preset from the list, build targets, debug integration, etc are provided.

You still need the above settings at some point, but only when you need more advanced features.

```
{
    "version": 4,
    "configurePresets": [
    {
      "name": "fprime",
      "displayName": "F´ Release Preset",
      "description": "F´ release build using local fprime-venv",
      "binaryDir": "${sourceDir}/build-fprime-automatic-native",
      "environment": {
        "VIRTUAL_ENV": "${fileDir}/fprime-venv",
        "PATH": "$env{VIRTUAL_ENV}/bin:$penv{PATH}"
      },
      "cacheVariables": {
          "CMAKE_EXPORT_COMPILE_COMMANDS": "ON",
          "CMAKE_BUILD_TYPE": "Release"
      }
    },
    {
      "name": "fprime-debug",
      "inherits": "fprime",
      "displayName": "F´ Debug Preset",
      "description": "F´ debug build using local fprime-venv",
      "cacheVariables": {
          "CMAKE_BUILD_TYPE": "Debug"
      }
    },
    {
      "name": "fprime-ut",
      "inherits": "fprime-debug",
      "displayName": "F´ Unit Test Preset",
      "description": "F´ debug build including unit tests using local fprime-venv",
      "binaryDir": "${sourceDir}/build-fprime-automatic-native-ut",
      "cacheVariables": {
          "BUILD_TESTING": "ON"
      }
    },
      {
        "name": "fprime-ninja",
        "displayName": "F´ Release (Ninja) Preset",
        "description": "F´ release build using ninja and local fprime-venv",
        "generator": "Ninja"
      },
      {
        "name": "fprime-debug-ninja",
        "inherits": "fprime-ninja",
        "displayName": "F´ Debug (Ninja) Preset",
        "description": "F´ debug build using ninja and local fprime-venv",
        "generator": "Ninja"
      },
      {
        "name": "fprime-ut-ninja",
        "inherits": "fprime-debug-ninja",
        "displayName": "F´ Unit Test (Ninja) Preset",
        "description": "F´ debug build including unit tests using ninja and local fprime-venv",
        "binaryDir": "${sourceDir}/build-fprime-automatic-native-ut",
        "generator": "Ninja"
      }
    ]
}
```

**Reply by LeStarch** [reply]
We will be shipping this file in `fprime-bootstrap` automatically to make development easier!

**Comment by Gonta01** [comment]
@thomas-bc and @LeStarch very grateful for your responses but to be honest it is a little bit difficult to catch for newbie(like me). I'll be very grateful for step by step tutorial for VS code configuration based on some other tutorial (for example [Hello world](https://fprime-community.github.io/fprime-tutorial-hello-world/))
